### ISSUE TITLE
FR: Ensure batch error responses can be matched to their causing message

### ISSUE THREAD

**Issue by GFriedrich** [firebase/firebase-admin-java#485]
If I'm not mistaken it is currently not possible to match the single responses of a `BatchResponse` to the very specific messages of the call to `sendAll` in `FirebaseMessagingClientImpl`.
This is due to the fact that responses could be redirected/retried and then the order of the responses will be mixed up in the list of the `BatchResponse` resulting in a potential wrong matching to their original message.
I would therefore appreciate, if it would be possible to somehow match the potential error response of a single message to their root causing message.
A potential reason is, to figure out a specific device token that might be outdated and therefore resulting in a failed request within the batch request.
Feel free to correct me, if I got anything wrong.
Thanks in advance.

**Comment by hiranya911** [comment]
The order of the responses is guaranteed to match the order of the input messages. From the [documentation](https://firebase.google.com/docs/reference/admin/java/reference/com/google/firebase/messaging/FirebaseMessaging#sendAll(java.util.List%3Ccom.google.firebase.messaging.Message%3E)):

> The responses list obtained by calling getResponses() on the return value corresponds to the order of input messages.

Retries doesn't work at individual message level. It's all or nothing.

**Comment by GFriedrich** [comment]
@hiranya911: Ok I was just confused due to the following facts:
If I check the implementation of the BatchUnparsedResponse:
https://github.com/googleapis/google-api-java-client/blob/0f6f940b70b4858b205424a2291820230b19d6e4/google-api-client/src/main/java/com/google/api/client/googleapis/batch/BatchUnparsedResponse.java#L190-L204
The retryAllowed property is true due to the outer batch request being allowed to be retried. (Even though the childRequestFactory doesn't support retries.)
This in turn will set the retrySupported to true.
As said before the childRequestFactory doesn't support retries so unsuccessfulResponseHandler is null.
This results in a call to handleRedirect.

I was now unsure whether this call could result for some of the responses in a "true" value, potentially resulting in those requests being added to the unsuccessfulRequestInfos list.
This of course won't happen if it is ensured that single responses never contain a result that may trigger that behaviour. And I guess this is the case, right?

Feel free to correct me, if I got anything wrong here.

**Comment by hiranya911** [comment]
Thanks @GFriedrich for the detailed and thorough analysis. We certainly don't want `BatchUnparsedResponse` to retry individual messages. To make sure this is the case I debugged through the following test case:

https://github.com/firebase/firebase-admin-java/blob/408b85f7d0072e8fb3b462aa2317843d683e7ad7/src/test/java/com/google/firebase/messaging/FirebaseMessagingClientImplTest.java#L372-L382

In `BatchUnparsedResponse` I observed that:
* `unsuccessfulResponseHandler` is null on child requests. Therefore `errorHandled` never gets set to true.
* FCM backend API doesn't send redirect status codes so `redirectRequest` doesn't get set to true.

These in turns cause the following condition to get skipped:

```java
if (retrySupported && (errorHandled || redirectRequest)) {
    unsuccessfulRequestInfos.add(requestInfo);
}
```

However, now that I've taken a close look at the source, I'd like to make this behavior more explicit. Easiest would be to do the following on the child requests:

```java
request.setUnsuccessfulResponseHandler(null);
request.setFollowRedirects(false);
```

I can create a PR to that effect sometime in the next few days.
